### PR TITLE
[GUI] Fix focus behaviour when scrolling to the bottom of the panel

### LINKED ISSUE #15626: Scroll to bottom of panel control using scrollbar wraps focus to first row if no item in focus position.
### Describe the bug
Here is a clear and concise description of what the problem is:

For panel controls, if you scroll to the bottom of the panel control using a scroll bar and there is no item in the focus position column, then focus wraps to the top row.

This is a problem not only because it shows info for an item that is not on screen, but also when you move off the scrollbar the focus is in a position which is off screen.

## Expected Behavior
Here is a clear and concise description of what was expected to happen:

If there is no item at the focus column position, focus should be on the last item in the panel rather than wrapping around.

## Actual Behavior

The focus wraps around to the top row.

### To Reproduce

1. Use info wall view in Estuary.
2. Find a folder where the number of items leaves a blank space in the bottom right of the panel.
3. Focus the bottom right position of the panel.
4. Move to the scrollbar and the scroll down the panel. 
5. Because there is no item in the very last position of the panel where focus was previously, the focus wraps around to the top row.

### Screenshots 
Here are some links or screenshots to help explain the problem:

In this screenshot you can see that as there is no item in the last row and column, the focus has wrapped around to the top row of the panel and the info is showing for "8 out of 10 cats" rather than "That Mitchell and Webb Look"

screenshot186

### Your Environment
Used Operating system:

 - [ ] Android
 - [ ] iOS
 - [ ] Linux
 - [ ] OSX
 - [ ] Raspberry-Pi
 - [x] Windows
 - [ ] Windows UWP

 - Operating system version/name: Windows 10
 - Kodi version: 18.1

*note: Once the issue is made we require you to update it with new information or Kodi versions should that be required.
Team Kodi will consider your problem report however, we will not make any promises the problem will be solved.*

### PR DESCRIPTION
**Pull request by joseluismarti**
## Description
As described in the issue #15626 if you scroll to the bottom of the panel using the scroll bar and there is no item near the focus, an item from the top row is selected. Displays the info for an item that is not on the screen.

This issue can also be reproduced by pointing the mouse at the bottom where there are no elements.

I propose to solve this by preventing the cursor from taking values that exceed the number of items being displayed. In these cases the cursor will be positioned on the last item.

## Types of change
- [X] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **Student submission** (PR was done for educational purposes and will be treated as such)
- [ ] **None of the above** (please explain below)

## Checklist:
- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

File: xbmc/guilib/GUIPanelContainer.cpp [hunk 1]
```diff
@@ -399,9 +399,17 @@ void CGUIPanelContainer::ValidateOffset()
...
 void CGUIPanelContainer::SetCursor(int cursor)
 {
+  // exceeds the number of items the panel can hold
   if (cursor > m_itemsPerPage * m_itemsPerRow - 1)
     cursor = m_itemsPerPage * m_itemsPerRow - 1;
-  if (cursor < 0) cursor = 0;
+
+  // exceeds the number of items being displayed
+  if (cursor > (int)m_items.size() - 1 - GetOffset() * m_itemsPerRow)
+    cursor = (int)m_items.size() - 1 - GetOffset() * m_itemsPerRow;
```
Comment (ksooo) [inline xbmc/guilib/GUIPanelContainer.cpp:408]: Please, no c-style casts. Coding Guideline!

Comment (joseluismarti) [reply xbmc/guilib/GUIPanelContainer.cpp:408]: You're right, thanks

**Comment by joseluismarti** [comment]
If "Show parent folder items" is enabled, shows the first item at the top "<-", its info is ".."

If disable, shows the info of the item that is first at the top.

It makes no difference to the fix.

To better understand the issue, leave only one item at the bottom and point the mouse to the right along the other "3 columns" where there is no item. It will show the info of the items in the first row.

**Comment by HitcherUK** [comment]
Windows test build [here](http://mirrors.kodi.tv/test-builds/windows/win64/KodiSetup-20231214-41529b9c-video-version-use-standard-select-dialog-x64.exe).

**Comment by CrystalP** [comment]
Oh yes, it wraps. If there is more than one blank spots at the bottom, first blank shows first item of the list, 2nd blank shows 2nd item, ...

**Comment by CrystalP** [comment]
Fix works fine and located in a reasonable place. No surprises with actions like page up/down or list with single item.

**Comment by jurialmunkey** [comment]
Working well for me with testbuild above.

Tested with/without parent folder and also "*All Items" in library. Also tested a custom panel with `browse="always"` for the more item at the end and works well there too.

**Comment by jurialmunkey** [comment]
Actually, managed to get some weird behaviour with the "More" item active.

See video below.

Info for the corresponding item from the top row is briefly displayed before the more item gains focus. You can see that with the fanart on the left briefly changing in the video:


https://github.com/xbmc/xbmc/assets/4618719/48fb7211-be80-4f6c-ae68-676aa6786bfc


Here's a gist with my test code in a custom window. I reappropriated the AddonLauncher custom window just to make it easy to get to

https://gist.github.com/jurialmunkey/22f454bd8deb8385360ad05dbecd3376

File: xbmc/guilib/GUIPanelContainer.cpp [hunk 2]
```diff
@@ -399,9 +399,17 @@ void CGUIPanelContainer::ValidateOffset()
 
 void CGUIPanelContainer::SetCursor(int cursor)
 {
+  // exceeds the number of items the panel can hold
   if (cursor > m_itemsPerPage * m_itemsPerRow - 1)
     cursor = m_itemsPerPage * m_itemsPerRow - 1;
-  if (cursor < 0) cursor = 0;
+
+  // exceeds the number of items being displayed
+  if (cursor > static_cast<int>(m_items.size()) - 1 - GetOffset() * m_itemsPerRow)
```
Comment (ksooo) [inline xbmc/guilib/GUIPanelContainer.cpp:407]: You are doing the same calculation two times in a row. Don’t you think you should calculate once and store the result in a variable?

Comment (joseluismarti) [reply xbmc/guilib/GUIPanelContainer.cpp:407]: Right, initially while testing I was using a temporary variable, but in the end I changed it. I will change it

**Comment by joseluismarti** [comment]
In the video we can see how when scrolling there is a moment when shows the info of an item at the top. It's a similar issue to the one I'm trying to correct with this PR. I've done some tests but I haven't found the right place to adjust the cursor and the focus, I would have to spend more time to analyse the cause.

I think we can leave this PR as it is. Leaving this new issue for a next PR when we find the cause.

**Comment by jurialmunkey** [comment]
> I think we can leave this PR as it is. Leaving this new issue for a next PR when we find the cause.

@joseluismarti - I agree. Not worth blocking this PR over. Just thought I'd mention it since I spotted it.

Overall behaviour is much improved with this PR. Even if it has to travel over the top row first, it still lands on the "correct" item eventually and that's a considerable improvement compared to the old behaviour where focus ended up offscreen.

Appreciate you taking the time to look at this!

**Comment by CrystalP** [comment]
@jurialmunkey's problem happens because the scrolling offset is set immediately to the new value as the scrolling starts, but the function you modified is called when the scrolling animation ends (ok for mouse though).

This seems to work to reevaluate the cursor position when scrolling starts:
```c++
diff --git a/xbmc/guilib/GUIBaseContainer.cpp b/xbmc/guilib/GUIBaseContainer.cpp
index 3db7db9364e..5f3b81abfbd 100644
--- a/xbmc/guilib/GUIBaseContainer.cpp
+++ b/xbmc/guilib/GUIBaseContainer.cpp
@@ -516,6 +516,7 @@ bool CGUIBaseContainer::OnMessage(CGUIMessage& message)
         if (message.GetParam1() != GetOffset())
           m_pageChangeTimer.StartZero();
         ScrollToOffset(message.GetParam1());
+        SetCursor(GetCursor());
         return true;
       }
     }
```

I'm just not sure what the `SetContainerMoving()` in  `CGUIPanelContainer::SetCursor()` does and if it could cause problem if the container is scrolling.

**Comment by joseluismarti** [comment]
I've tried the change you propose and still shows a moment an item at the top when scrolling, we'll keep looking for a fix.

**Comment by CrystalP** [comment]
Yes I could still see something with a larger list and scrolling with the page down key from a specific position.
This is in a better location ee6941b0c61a1ecac9b762c0d356cc4ec9a27bcc

But not a finished version, there is overlap with what you did, probably a new void CGUIPanelContainer::ValidateCursor() function extracting the logic to override cursor value too high is needed to clean up.

**Comment by joseluismarti** [comment]
> This is in a better location [ee6941b](https://github.com/xbmc/xbmc/commit/ee6941b0c61a1ecac9b762c0d356cc4ec9a27bcc)

I've tested this change and it seems to work fine

**Comment by joseluismarti** [comment]
> Overall behaviour is much improved with this PR. Even if it has to travel over the top row first, it still lands on the "correct" item eventually and that's a considerable improvement compared to the old behaviour where focus ended up offscreen.

This has been left here, I think it can be merged

**Comment by CrystalP** [comment]
Would you put the finishing touch on my patch? Then there'd be no need to revisit later.

**Comment by joseluismarti** [comment]
Sure, I thought something else was missing, I'll add it.

**Comment by CrystalP** [comment]
OK but it looks like I'm not communicating properly with you, I posted the patch with a note

> not a finished version, there is overlap with what you did, probably a new void CGUIPanelContainer::ValidateCursor() function extracting the logic to override cursor value too high is needed to clean up.

The new ValidateCursor() is what I was referring to as a small refactor to finish up the PR. Is that clearer?

**Comment by joseluismarti** [comment]
Do you mean a `CGUIPanelContainer::ValidateCursor()` function that contains all the cursor settings that are now made in `SetCursor()` and returns the valid cursor value?

If you refer to this refact I see well to leave the cursor adjustment in the `SetCursor()` function itself, the code is short and clear, and we would only call to this function from `SetCursor()`.

If not, I didn't understand you :)

**Comment by CrystalP** [comment]
That was it more or less but I looked into it a bit more and it's not necessary.

**Review by CrystalP — APPROVED** [review]
Fixes a weird behavior for the last items of a panel.

**Pull request merged**